patches/api/0439-Provide-registered-listeners-in-the-PluginManager.patch

```diff
@@ -0,0 +1,215 @@
++            listener.unregisterHandlerList(this); // Paper - faster registration
+             handlers = null;
+         }
+     }
+@@ -159,7 +166,11 @@ public class HandlerList {
+         boolean changed = false;
+         for (List<RegisteredListener> list : handlerslots.values()) {
+             for (ListIterator<RegisteredListener> i = list.listIterator(); i.hasNext();) {
+-                if (i.next().getPlugin().equals(plugin)) {
++                // Paper start - faster registration
++                var listener = i.next();
++                if (listener.getPlugin().equals(plugin)) {
++                    listener.unregisterHandlerList(this);
++                    // Paper end
+                     i.remove();
+                     changed = true;
+                 }
+@@ -177,7 +188,11 @@ public class HandlerList {
+         boolean changed = false;
+         for (List<RegisteredListener> list : handlerslots.values()) {
+             for (ListIterator<RegisteredListener> i = list.listIterator(); i.hasNext();) {
+-                if (i.next().getListener().equals(listener)) {
++                // Paper start - faster registration
++                var j = i.next();
++                if (j.getListener().equals(listener)) {
++                    j.unregisterHandlerList(this);
++                    // Paper end
+                     i.remove();
+                     changed = true;
+                 }
+diff --git a/src/main/java/org/bukkit/plugin/PluginManager.java b/src/main/java/org/bukkit/plugin/PluginManager.java
+index 03213fde8315384ec56c16031cfc606ade2e8091..33b322078d568e174eadd424463abf745a171d12 100644
+--- a/src/main/java/org/bukkit/plugin/PluginManager.java
++++ b/src/main/java/org/bukkit/plugin/PluginManager.java
+@@ -118,6 +118,7 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
+      * @param listener Listener to register
+      * @param plugin Plugin to register
+      */
++    @Deprecated(forRemoval = true) // Paper - superseded by registerListeners
+     public void registerEvents(@NotNull Listener listener, @NotNull Plugin plugin);
+ 
+     /**
+@@ -129,6 +130,7 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
+      * @param executor EventExecutor to register
+      * @param plugin Plugin to register
+      */
++    @Deprecated(forRemoval = true) // Paper - superseded by registerListener
+     public void registerEvent(@NotNull Class<? extends Event> event, @NotNull Listener listener, @NotNull EventPriority priority, @NotNull EventExecutor executor, @NotNull Plugin plugin);
+ 
+     /**
+@@ -141,6 +143,7 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
+      * @param plugin Plugin to register
+      * @param ignoreCancelled Whether to pass cancelled events or not
+      */
++    @Deprecated(forRemoval = true) // Paper - superseded by registerListener
```

Comment:
I think these deprecations are going too far. I don't ever see us removing these since they work fine. I think they are a better candidate for the "Obsolete" annotation.

Reply:
Is there an annotation for Obsolete or should I just only note its obsoleteness in the lore?

Reply:
There is, it’s “ApiStatus.Obsolete” from jetbrains’ annotation library

Reply:
Marking the current methods as obsolete just because of the return value seems a bit "strict" to me, imo the new methods could just be referenced in a seeAlso javadoc comment